### PR TITLE
Let kwargs_to_strings work with default values and postional arguments

### PR DESCRIPTION
**Pull request by seisman**
**Description of proposed changes**

See https://github.com/GenericMappingTools/pygmt/issues/2361 for the original issue report. The solution comes from https://stackoverflow.com/a/69170441. 

In summary, it's possible to use the following codes to get the default values of parameters:
```
sig = inspect.signature(f)
bound = sig.bind(*args, **kwds)
bound.apply_defaults()
```

Here are some example codes to understand the behavior the the above codes, in which case 3 is usually how we call PyGMT functions/methods:
```python
import inspect

def func(x, y, z=3, **kwargs):
    """
    An example function which 
    """
    print(f"x={x} y={y} z={z}")
    print(kwargs)

print("Case 1:")
sig = inspect.signature(func)
bound = sig.bind(1, 2)
print(bound.arguments)  # raw arguments
bound.apply_defaults()
print(bound.arguments)  # after apply_defaults

print("Case 2:")
bound = sig.bind(1, 2, 4, R=[1, 2, 3, 4], J="X10c")
bound.apply_defaults()
print(bound.arguments)

print("Case 3:")
bound = sig.bind(1, 2, z=4, R=[1, 2, 3, 4], J="X10c")
bound.apply_defaults()
print(bound.arguments)

bound.arguments["kwargs"]["R"] = "1/2/3/4"
print(bound.arguments)
print(bound.args, bound.kwargs)

func(*bound.args, **bound.kwargs)

```
The outputs are:
```
Case 1:
{'x': 1, 'y': 2}
{'x': 1, 'y': 2, 'z': 3, 'kwargs': {}}
Case 2:
{'x': 1, 'y': 2, 'z': 4, 'kwargs': {'R': [1, 2, 3, 4], 'J': 'X10c'}}
Case 3:
{'x': 1, 'y': 2, 'z': 4, 'kwargs': {'R': [1, 2, 3, 4], 'J': 'X10c'}}
{'x': 1, 'y': 2, 'z': 4, 'kwargs': {'R': '1/2/3/4', 'J': 'X10c'}}
(1, 2, 4) {'R': '1/2/3/4', 'J': 'X10c'}
x=1 y=2 z=4
{'R': '1/2/3/4', 'J': 'X10c'}
```

This PR will fix https://github.com/GenericMappingTools/pygmt/issues/2361. With changes in this PR, now we don't need to manually deal with `offset` in `Figure.timestamp` (see 
https://github.com/GenericMappingTools/pygmt/pull/2208), `panel` in `subplot` (3edb1eebf2f7ec8bda9483c86c30484eb998720c) and `fname` in `pygmt.which` (https://github.com/GenericMappingTools/pygmt/pull/2726).

**Reminders**

- [x] Run `make format` and `make check` to make sure the code follows the style guide.
- [x] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Add new public functions/methods/classes to `doc/api/index.rst`.
- [ ] Write detailed docstrings for all functions/methods.
- [ ] If wrapping a new module, open a 'Wrap new GMT module' issue and submit reasonably-sized PRs.
- [ ] If adding new functionality, add an example to docstrings or tutorials.
- [ ] Use underscores (not hyphens) in names of Python files and directories.

**Slash Commands**

You can write slash commands (`/command`) in the first line of a comment to perform
specific operations. Supported slash commands are:

- `/format`: automatically format and lint the code
- `/test-gmt-dev`: run full tests on the latest GMT development version

### CODE REVIEW

File: pygmt/helpers/decorators.py [hunk 1]
```diff
@@ -686,58 +685,97 @@ def kwargs_to_strings(**conversions):
...
+                elif arg in bound.arguments.get("kwargs"):
+                    value = bound.arguments["kwargs"][arg]
+                else:
+                    continue
+
+                issequence = fmt in separators
+                if issequence and is_nonstr_iter(value):
+                    for index, item in enumerate(value):
+                        try:
+                            # check if there is a space " " when converting
```
Comment (michaelgrund) [inline pygmt/helpers/decorators.py:759]:
```suggestion
                            # Check if there is a space " " when converting
```

File: pygmt/helpers/decorators.py [hunk 2]
```diff
@@ -686,58 +685,97 @@ def kwargs_to_strings(**conversions):
...
+                            # a pandas.Timestamp/xr.DataArray to a string.
+                            # If so, use np.datetime_as_string instead.
+                            assert " " not in str(item)
+                        except AssertionError:
+                            # convert datetime-like item to ISO 8601
```
Comment (michaelgrund) [inline pygmt/helpers/decorators.py:764]:
```suggestion
                            # Convert datetime-like item to ISO 8601
```

File: pygmt/helpers/decorators.py [hunk 3]
```diff
@@ -686,58 +685,97 @@ def kwargs_to_strings(**conversions):
...
+                            # string format like YYYY-MM-DDThh:mm:ss.ffffff
```
Comment (michaelgrund) [inline pygmt/helpers/decorators.py:765]:
```suggestion
                            # string format like YYYY-MM-DDThh:mm:ss.ffffff.
```

**Comment by seisman** [comment]
I've merged this PR without review and approval so that I can work on other issues that need the changes in this PR.

**Pull request merged**